File: src/tests/Behavioral/MapArrayValueZero/MapArrayValueZero.go
// A map READ that MISSES yields the element type's Go zero value, and for a fixed-size array that
// zero is [N]T with N zeroed elements — not an empty array. `array<T>` carries its Go length in the
// instance, so `default(array<T>)` has LENGTH ZERO, and the first index into a missed entry panicked
// `index out of range [0] with length 0` where Go reads a zero.
//
// The live witness was html's unescapeEntity: `entity2` is `map[string][2]rune` and
// `if x := entity2[string(entityName)]; x[0] != 0` is the NORMAL path for any `&…` run that is not a
// two-rune entity, so `TestUnescape` died on ordinary input (html 2/3).
//
// Every read form is covered here because the shape has to come from the READ SITE — the map cannot
// supply it (a nil map has no entries and no type-level length), so each form needs its own seat:
// plain read, comma-ok read, a NESTED element whose inner lengths must survive too, a read through a
// NAMED map type (the generated wrapper forwards the overload), and both maps that can never have
// the entry — nil and empty.
package main

import "fmt"

var entity2 = map[string][2]rune{
	"NotEqualTilde;": {'≂', '̸'},
}

// Plain read, hit and miss, plus the comma-ok form.
func mapMiss() {
	x := entity2["notthere"]
	fmt.Println("miss:", len(x), x[0], x[1])

	y := entity2["NotEqualTilde;"]
	fmt.Println("hit:", len(y), y[0], y[1])

	z, ok := entity2["alsomissing"]
	fmt.Println("commaok:", ok, len(z), z[0], z[1])

	w, ok2 := entity2["NotEqualTilde;"]
	fmt.Println("commaok hit:", ok2, len(w), w[0], w[1])
}

// A NESTED element: the inner arrays are zero-length too unless the element factory rides along.
var nested = map[string][2][3]int{"a": {{1, 2, 3}, {4, 5, 6}}}

func mapMissNested() {
	n := nested["zzz"]
	fmt.Println("nested miss:", len(n), len(n[0]), len(n[1]), n[0][2], n[1][0])

	h := nested["a"]
	fmt.Println("nested hit:", len(h), len(h[0]), h[0][2], h[1][0])
}

// A NAMED map type reads through the generated wrapper rather than golib's map directly.
type quadMap map[int][4]byte

func namedMap() {
	var nilMap quadMap // nil: no store at all, so the zero can only come from the read site
	v := nilMap[7]
	fmt.Println("nil named:", len(v), v[0], v[3])

	empty := make(quadMap)
	e := empty[9]
	fmt.Println("empty named:", len(e), e[0], e[3])

	empty[9] = [4]byte{'a', 'b', 'c', 'd'}
	got, ok := empty[9]
	fmt.Println("named hit:", ok, len(got), got[0], got[3])
}

// An UNNAMED map local, both nil and empty, to prove the read site (not the construction) carries it.
func plainNilAndEmpty() {
	var nilMap map[string][2]rune
	a := nilMap["x"]
	fmt.Println("nil plain:", len(a), a[0], a[1])

	emptyMap := map[string][2]rune{}
	b := emptyMap["x"]
	fmt.Println("empty plain:", len(b), b[0], b[1])
}

// A STORE is untouched: it has a value and needs no zero.
func storeThenRead() {
	m := make(map[string][2]rune)
	m["k"] = [2]rune{'q', 'r'}
	fmt.Println("stored:", len(m["k"]), m["k"][0], m["k"][1])
	fmt.Println("other:", len(m["other"]), m["other"][0])
}

func main() {
	mapMiss()
	mapMissNested()
	namedMap()
	plainNilAndEmpty()
	storeThenRead()
}

File: src/tests/Behavioral/MapArrayValueZero/MapArrayValueZero.cs
namespace go;

using fmt = fmt_package;

partial class main_package {

internal static map<@string, array<rune>> entity2 = new map<@string, array<rune>>{
    ["NotEqualTilde;"u8] = new rune[]{(rune)'≂', (rune)'̸'}.array()
};

// Hoisted @string literals (single allocation; Go keeps these in RODATA)
private static readonly @string notthereˢ = "notthere"u8;
private static readonly object missˢ = (@string)"miss:"u8;
private static readonly @string notEqualTildeˢ = "NotEqualTilde;"u8;
private static readonly object hitˢ = (@string)"hit:"u8;
private static readonly @string alsomissingˢ = "alsomissing"u8;
private static readonly object commaokˢ = (@string)"commaok:"u8;
private static readonly object commaokHitˢ = (@string)"commaok hit:"u8;

internal static void mapMiss() {
    var x = entity2[notthereˢ, () => new array<rune>(2)].Clone();
    fmt.Println(missˢ, len(x), x[0], x[1]);
    var y = entity2[notEqualTildeˢ, () => new array<rune>(2)].Clone();
    fmt.Println(hitˢ, len(y), y[0], y[1]);
    var (z, ok) = entity2[alsomissingˢ, () => new array<rune>(2), ꟷ];
    fmt.Println(commaokˢ, ok, len(z), z[0], z[1]);
    var (w, ok2) = entity2[notEqualTildeˢ, () => new array<rune>(2), ꟷ];
    fmt.Println(commaokHitˢ, ok2, len(w), w[0], w[1]);
}

internal static map<@string, array<array<nint>>> nested = new map<@string, array<array<nint>>>{["a"u8] = new array<nint>[]{new nint[]{1, 2, 3}.array(), new nint[]{4, 5, 6}.array()}.array()};

// Hoisted @string literals (single allocation; Go keeps these in RODATA)
private static readonly @string zzzˢ = "zzz"u8;
private static readonly object nestedMissˢ = (@string)"nested miss:"u8;
private static readonly object nestedHitˢ = (@string)"nested hit:"u8;

internal static void mapMissNested() {
    var n = nested[zzzˢ, () => new array<array<nint>>(2, () => new(3))].Clone();
    fmt.Println(nestedMissˢ, len(n), len(n[0]), len(n[1]), n[0][2], n[1][0]);
    var h = nested["a"u8, () => new array<array<nint>>(2, () => new(3))].Clone();
    fmt.Println(nestedHitˢ, len(h), len(h[0]), h[0][2], h[1][0]);
}

[GoType("map[nint, array<byte>]")] partial struct quadMap;

// Hoisted @string literals (single allocation; Go keeps these in RODATA)
private static readonly object nilNamedˢ = (@string)"nil named:"u8;
private static readonly object emptyNamedˢ = (@string)"empty named:"u8;
private static readonly object namedHitˢ = (@string)"named hit:"u8;

internal static void namedMap() {
    quadMap nilMap = default!;
    var v = nilMap[7, () => new array<byte>(4)].Clone();
    fmt.Println(nilNamedˢ, len(v), v[0], v[3]);
    var empty = new quadMap(0);
    var e = empty[9, () => new array<byte>(4)].Clone();
    fmt.Println(emptyNamedˢ, len(e), e[0], e[3]);
    empty[9] = new byte[]{(rune)'a', (rune)'b', (rune)'c', (rune)'d'}.array();
    var (got, ok) = empty[9, () => new array<byte>(4), ꟷ];
    fmt.Println(namedHitˢ, ok, len(got), got[0], got[3]);
}

// Hoisted @string literals (single allocation; Go keeps these in RODATA)
private static readonly object nilPlainˢ = (@string)"nil plain:"u8;
private static readonly object emptyPlainˢ = (@string)"empty plain:"u8;

internal static void plainNilAndEmpty() {
    map<@string, array<rune>> nilMap = default!;
    var a = nilMap["x"u8, () => new array<rune>(2)].Clone();
    fmt.Println(nilPlainˢ, len(a), a[0], a[1]);
    var emptyMap = new map<@string, array<rune>>{};
    var b = emptyMap["x"u8, () => new array<rune>(2)].Clone();
    fmt.Println(emptyPlainˢ, len(b), b[0], b[1]);
}

// Hoisted @string literals (single allocation; Go keeps these in RODATA)
private static readonly object storedˢ = (@string)"stored:"u8;
private static readonly object otherˢ = (@string)"other:"u8;
private static readonly @string otherˢ2 = "other"u8;

internal static void storeThenRead() {
    var m = new map<@string, array<rune>>();
    m["k"u8] = new rune[]{(rune)'q', (rune)'r'}.array();
    fmt.Println(storedˢ, len(m["k"u8, () => new array<rune>(2)]), m["k"u8, () => new array<rune>(2)][0], m["k"u8, () => new array<rune>(2)][1]);
    fmt.Println(otherˢ, len(m[otherˢ2, () => new array<rune>(2)]), m[otherˢ2, () => new array<rune>(2)][0]);
}

internal static void Main() {
    mapMiss();
    mapMissNested();
    namedMap();
    plainNilAndEmpty();
    storeThenRead();
}

} // end main_package
